### PR TITLE
Mc/ix toggler

### PR DESCRIPTION
**Pull request by microwavedcola1**

### CODE REVIEW

File: programs/mango-v4/src/error.rs [hunk 1]
```diff
@@ -69,8 +69,6 @@ pub enum MangoError {
     TokenInReduceOnlyMode,
     #[msg("market is in reduce only mode")]
     MarketInReduceOnlyMode,
-    #[msg("group is halted")]
-    GroupIsHalted,
```
Comment (ckamm) [inline programs/mango-v4/src/error.rs:73]: Errors are sequentially numbered: can't ever remove any, can only rename

File: programs/mango-v4/src/error.rs [hunk 2]
```diff
@@ -89,6 +87,8 @@ pub enum MangoError {
     HasOpenPerpTakerFills,
     #[msg("deposit crosses the current group deposit limit")]
     DepositLimit,
+    #[msg("ix is disabled")]
```
Comment (ckamm) [inline programs/mango-v4/src/error.rs:90]: instruction

File: programs/mango-v4/src/instructions/ix_gate_set.rs [hunk 3]
```diff
@@ -0,0 +1,20 @@
+use anchor_lang::prelude::*;
+
+use crate::state::*;
+
+#[derive(Accounts)]
+pub struct IxGateSet<'info> {
+    #[account(
+        mut,
+        has_one = admin,
```
Comment (ckamm) [inline programs/mango-v4/src/instructions/ix_gate_set.rs:9]: should be available to the security admin (that is, disabling instructions should be available, not re-enabling)

File: programs/mango-v4/src/state/group.rs [hunk 4]
```diff
@@ -69,11 +71,75 @@ impl Group {
...
+/// - Benchmark,
+/// - ComputeAccountData,
+/// - GroupCreate
+/// - GroupEdit
+/// - HealthRegion
+/// - IxGateSet,
+/// - PerpZeroOut,
+/// - PerpEditMarket,
+/// - TokenEdit,
+pub enum IxGate {
```
Comment (ckamm) [inline programs/mango-v4/src/state/group.rs:93]: suggest adding repr(u128)

File: programs/mango-v4/src/state/group.rs [hunk 5]
```diff
@@ -69,11 +71,75 @@ impl Group {
...
+/// Enum for lookup into ix gate
+/// note:
+/// total ix files 56,
+/// ix files included 47,
+/// ix files not included 9,
+/// - Benchmark,
+/// - ComputeAccountData,
+/// - GroupCreate
+/// - GroupEdit
+/// - HealthRegion
```
Comment (ckamm) [inline programs/mango-v4/src/state/group.rs:88]: would be good to also be able to disable HealthRegion

File: programs/mango-v4/src/state/group.rs [hunk 6]
```diff
@@ -69,11 +71,75 @@ impl Group {
...
+/// - IxGateSet,
+/// - PerpZeroOut,
+/// - PerpEditMarket,
+/// - TokenEdit,
+pub enum IxGate {
+    AccountClose = 0,
+    AccountCreate,
```
Comment (ckamm) [inline programs/mango-v4/src/state/group.rs:95]: I'd even suggest explicitly numbering all enum variants, so it's impossible to accidentally reuse/reshuffle numbers

File: programs/mango-v4/tests/test_ix_gate_set.rs [hunk 7]
```diff
@@ -0,0 +1,120 @@
...
+            reduce_only: false,
+            account,
+            owner,
+            token_account: payer_mint0_account,
+            token_authority: payer.clone(),
+            bank_index: 0,
+        },
+    )
+    .await
+    .unwrap();
```
Comment (ckamm) [inline programs/mango-v4/tests/test_ix_gate_set.rs:67]: could be replaced by create_funded_account()

File: ts/client/src/clientIxParamBuilder.ts [hunk 8]
```diff
@@ -110,3 +111,173 @@ export const NullPerpEditParams: PerpEditParams = {
...
+  ixGate.ior(p.TokenAddBank ? new BN(0) : new BN(1).ushln(38));
+  ixGate.ior(p.TokenDeposit ? new BN(0) : new BN(1).ushln(39));
+
+  ixGate.ior(p.TokenDeregister ? new BN(0) : new BN(1).ushln(40));
+  ixGate.ior(p.TokenLiqBankruptcy ? new BN(0) : new BN(1).ushln(41));
+  ixGate.ior(p.TokenLiqWithToken ? new BN(0) : new BN(1).ushln(42));
+  ixGate.ior(p.TokenRegister ? new BN(0) : new BN(1).ushln(43));
+  ixGate.ior(p.TokenRegisterTrustless ? new BN(0) : new BN(1).ushln(44));
+  ixGate.ior(p.TokenUpdateIndexAndRate ? new BN(0) : new BN(1).ushln(45));
+  ixGate.ior(p.TokenWithdraw ? new BN(0) : new BN(1).ushln(46));
```
Comment (ckamm) [inline ts/client/src/clientIxParamBuilder.ts:280]: this was very annoying to verify, is there a way to remove the repetition? (optional)

File: programs/mango-v4/src/state/group.rs (hunk 4 above)
Comment (microwavedcola1) [reply programs/mango-v4/src/state/group.rs:93]: did you see this https://github.com/rust-lang/rust/issues/56071

File: programs/mango-v4/tests/program_test/mango_client.rs [hunk 9]
```diff
@@ -3629,6 +3628,7 @@ impl ClientInstruction for ComputeAccountDataInstruction {
 }
 
 pub struct HealthRegionBeginInstruction {
+    pub group: Pubkey,
     pub account: Pubkey,
```
Comment (ckamm) [inline programs/mango-v4/tests/program_test/mango_client.rs:3632]: no need to require clients to pass in the group, just read it from the account

File: programs/mango-v4/tests/test_ix_gate_set.rs [hunk 10]
```diff
@@ -34,22 +33,17 @@ async fn test_ix_gate_set() -> Result<(), TransportError> {
...
+        0,
+        &context.users[1],
+        &mints[0..1],
+        10,
+        0,
     )
-    .await
-    .unwrap()
-    .account;
+    .await;
```
Comment (ckamm) [inline programs/mango-v4/tests/test_ix_gate_set.rs:46]: you kept the explicit deposit that follows after this; probably that can be removed?

**Review by ckamm — COMMENTED** [review]
Could you add a test that disables all instructions, just to confirm it fits into cu budget?

**Pull request merged**